### DISCUSSION TITLE
Automatically categorise keywords

### DISCUSSION THREAD

**Discussion by hmltn-0** [explosion/spaCy#9921] [Help: Coding & Implementations] [answered]
I would like to extract keywords and automatically categorise them. The algorithm should guess how to group them together and what their category label should be from scratch, with no suggested keywords.

I know of some Spacy methods for keyword extraction but I don’t know if the need for auto-categorisation implies I can just hitch on one more step in the procedure or if I should use a different algorithm altogether.

I am pretty sure the algorithm needs to evaluate every keyword with relevance to the entire text. Maybe an algorithm can rank the similarity of each keyword to each other keyword (using BERT). Then another algorithm can “cluster” terms with high degrees of interlinking. Lastly, GPT-3 or BERT can suggest a label for each group of terms.

I’m trying to think of ways to combine how some algorithms are based on pure statistics and some are heavily pre-trained and some are trained on the specific data and combine them. Like training BERT on the text I’m using, doing a graph-theoretic clustering of key terms or a more linguistic clustering using semantic relations like WordNet, and either something like WordNet or trained BERT as said to suggest a category.

Is there a more elegant way to do this? What’s the most standard way to cluster any kind of data points, and could it be applied to language data?

Thank you

**Comment by polm** [comment, marked as answer]
Clustering data is basically its own research field, and there are many, many approaches. A simple one you can get started with is KNN. 

There are techniques to generate labels for clusters but that kind of NLG is out of scope for spaCy. 

About your approach more broadly here, it's not clear to me what you're trying to do or why this would be useful. Can you give a more concrete example? More practically speaking, while I can see how a human could do this, how would you say whether one solution was better or worse than another? If you can't tell the computer if it's doing a good job or not there's no real way to train a model for this.

**Comment by hmltn-0** [comment]
Thank you very much.

I’m attempting to make a glossary creation tool. Given source text as
input, generate a glossary of key terms. The terms should ideally be
categorized as well as include definitions and context sentences pulled
from the text showing their use.

I hear you that right and wrong is fuzzier here than in some cases, but I
don’t see that as a huge limitation. I do think it might hint at that
however I pull this off, it would be with something unsupervised to group
terms rather than any kind of training set highlighting right and wrong
answers.

Someone suggested “topic modelling” to me so I may explore that.

Thanks very much.
